src/Sarif.Multitool/Rules/SARIF1022.PreferMessageIdInsteadOfMessageText.cs

```diff
@@ -0,0 +1,108 @@
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class PreferMessageIdInsteadOfMessageText : SarifValidationSkimmerBase
+    {
+        private IList<ReportingDescriptor> currentRules;
+
+        private readonly MultiformatMessageString _fullDescription = new MultiformatMessageString
+        {
+            Text = RuleResources.SARIF1022_PreferMessageIdInsteadOfMessageText
+        };
+
+        public override MultiformatMessageString FullDescription => _fullDescription;
+
+        public override FailureLevel DefaultLevel => FailureLevel.Error;
+
+        /// <summary>
+        /// SARIF1019
+        /// </summary>
+        public override string Id => RuleId.PreferMessageIdInsteadOfMessageText;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF1022_RuleMetadataDoesntIncludeMessageId),
+            nameof(RuleResources.SARIF1022_RuleMetadataIsNotConsistentWithArguments),
+            nameof(RuleResources.SARIF1022_ResultShouldUseMessageIdInsteadOfMessageText)
+        };
+
+        protected override void Analyze(Run run, string runPointer)
+        {
+            this.currentRules = run?.Tool?.Driver?.Rules;
+        }
+
+        protected override void Analyze(Result result, string resultPointer)
+        {
+            // if message.text is not null, we have a problem
+            if (!string.IsNullOrEmpty(result.Message.Text))
+            {
+                LogResult(resultPointer, nameof(RuleResources.SARIF1022_ResultShouldUseMessageIdInsteadOfMessageText), result.Message.Text);
+                return;
+            }
+
+            // if message.id is not null, we have to check if that exists in rules
+            if (!string.IsNullOrEmpty(result.Message.Id))
+            {
+                // we dont have rules or we didnt find a match, we have a problem
+                if (this.currentRules == null || this.currentRules.Count == 0 || this.currentRules.All(r => !r.MessageStrings.ContainsKey(result.Message.Id)))
+                {
+                    LogResult(resultPointer, nameof(RuleResources.SARIF1022_RuleMetadataDoesntIncludeMessageId), result.Message.Id);
+                    return;
+                }
+
+                // we have rules, we find it but, the count of parameters are invalid
+                string messageText = this.currentRules
+                    .First(r => r.MessageStrings.ContainsKey(result.Message.Id))
+                    .MessageStrings[result.Message.Id].Text;
+                if (!IsPlacehodlerInTextValid(messageText, result.Message.Arguments?.Count ?? 0))
+                {
+                    LogResult(resultPointer, nameof(RuleResources.SARIF1022_RuleMetadataIsNotConsistentWithArguments), messageText, result.Message.Arguments.Count.ToString());
+                    return;
+                }
+            }
+        }
+
+        private bool IsPlacehodlerInTextValid(string text, int arguments)
```

Comment:
By the way, by "non-consecutive", I do _not_ mean that the placeholders have to appear in order.

This is fine: "{0}:  ... {1} ... {2}"
This is also fine: "{0}: ... {2} ... {1}"

But this is not fine: "{0}: ... {2}"
... because {1} is missing.

Even that last case is not _necessarily_ a problem. Perhaps the tool authors decided that the didn't want to display one of the arguments after all. But we _should_ warn about it, because it _might_ be a mistake.

---
In reply to: [434719411](https://github.com/microsoft/sarif-sdk/pull/1901#discussion_r434719411) [](ancestors = 434719411)

Reply:
just changed. let me know what do u think about the new methods.